demos/local-only-todolist/lib/powersync.dart

```diff
@@ -0,0 +1,219 @@
+        if (op.op == UpdateType.put) {
+          var data = Map<String, dynamic>.of(op.opData!);
+          data['id'] = op.id;
+          await table.upsert(data);
+        } else if (op.op == UpdateType.patch) {
+          await table.update(op.opData!).eq('id', op.id);
+        } else if (op.op == UpdateType.delete) {
+          await table.delete().eq('id', op.id);
+        }
+      }
+
+      // All operations successful.
+      await transaction.complete();
+    } on PostgrestException catch (e) {
+      if (e.code != null &&
+          fatalResponseCodes.any((re) => re.hasMatch(e.code!))) {
+        /// Instead of blocking the queue with these errors,
+        /// discard the (rest of the) transaction.
+        ///
+        /// Note that these errors typically indicate a bug in the application.
+        /// If protecting against data loss is important, save the failing records
+        /// elsewhere instead of discarding, and/or notify the user.
+        log.severe('Data upload error - discarding $lastOp', e);
+        await transaction.complete();
+      } else {
+        // Error may be retryable - e.g. network error or temporary server error.
+        // Throwing an error here causes this call to be retried after a delay.
+        rethrow;
+      }
+    }
+  }
+}
+
+/// Global reference to the database
+late PowerSyncDatabase db;
+
+bool isLoggedIn() {
+  return Supabase.instance.client.auth.currentSession?.accessToken != null;
+}
+
+/// id of the user currently logged in
+String getUserId() {
+  return Supabase.instance.client.auth.currentSession?.user.id ??
+      '00000000-0000-0000-0000-000000000000'; // default ID supplied for local-only mode
+}
+
+Future<String> getDatabasePath() async {
+  const dbFilename = 'powersync-demo-local-only.db';
+  // getApplicationSupportDirectory is not supported on Web
+  if (kIsWeb) {
+    return dbFilename;
+  }
+  final dir = await getApplicationSupportDirectory();
+  return join(dir.path, dbFilename);
+}
+
+Future<void> openDatabase() async {
+  await _openDatabase();
+
+  await loadSupabase();
+  if (isLoggedIn()) {
+    // If the user is already logged in, connect immediately.
+    // Otherwise, connect once logged via login/signup view.
+    await connectDatabase();
+  }
+}
+
+Future<void> _openDatabase() async {
+  var isSyncMode = await getSyncEnabled();
+  db = PowerSyncDatabase(
+      schema: makeSchema(synced: isSyncMode),
+      path: await getDatabasePath(),
+      logger: attachedLogger);
+  await db.initialize();
+}
+
+Future<void> connectDatabase() async {
+  if (!isLoggedIn()) {
+    log.severe("Can't connect database without being signed in");
+  }
+  SupabaseConnector? currentConnector;
+  var isSyncMode = await getSyncEnabled();
+
+  if (!isSyncMode) {
+    await switchToSyncedSchema(db, getUserId());
+    // Without closing and reopening the database the list pages breaks if there is no data in the tables when logging in/signing up.
```

Comment:
@stevensJourney @rkistner 
Before adding this `close/_openDb` snippet for switching to online from an **empty** local-only dataset, the list page's watch was broken.
I could see the data being synced down, but they weren't ending up in the watch stream even after the watch called got reinitialized. If I put a delay after the `await db.connect(connector: currentConnector);` on line 194 - the initial data would sync down, but even then any new data I created on the app or in supabase didn't trigger the watch with the latest data.

Is this workaround alright, or do we need to consider this a bug and not a limitation?